File: hellojava/src/main/java/com/fathzer/chess/MoveGenerators.java
package com.fathzer.chess;

import java.util.Arrays;
import java.util.List;

final class MoveGenerators {
    private static final int ROW_WIDTH = 10;

    private static final MoveBuilder WHITE_KING_MOVE_BUILDER = new KingMoveBuilder(true);
    private static final MoveBuilder WHITE_QUEEN_MOVE_BUILDER = new SliderMoveBuilder(new Direction[] { Direction.NORTH, Direction.SOUTH, Direction.EAST, Direction.WEST, Direction.NORTH_EAST, Direction.NORTH_WEST, Direction.SOUTH_EAST, Direction.SOUTH_WEST }, true);
    private static final MoveBuilder WHITE_ROOK_MOVE_BUILDER = new SliderMoveBuilder(new Direction[] { Direction.NORTH, Direction.SOUTH, Direction.EAST, Direction.WEST }, true);
    private static final MoveBuilder WHITE_BISHOP_MOVE_BUILDER = new SliderMoveBuilder(new Direction[] { Direction.NORTH_EAST, Direction.NORTH_WEST, Direction.SOUTH_EAST, Direction.SOUTH_WEST }, true);
    private static final MoveBuilder WHITE_KNIGHT_MOVE_BUILDER = new KnightMoveBuilder(true);
    private static final MoveBuilder WHITE_PAWN_MOVE_BUILDER = new PawnMoveBuilder(true);

    private static final MoveBuilder BLACK_KING_MOVE_BUILDER = new KingMoveBuilder(false);
    private static final MoveBuilder BLACK_QUEEN_MOVE_BUILDER = new SliderMoveBuilder(new Direction[] { Direction.NORTH, Direction.SOUTH, Direction.EAST, Direction.WEST, Direction.NORTH_EAST, Direction.NORTH_WEST, Direction.SOUTH_EAST, Direction.SOUTH_WEST }, false);
    private static final MoveBuilder BLACK_ROOK_MOVE_BUILDER = new SliderMoveBuilder(new Direction[] { Direction.NORTH, Direction.SOUTH, Direction.EAST, Direction.WEST }, false);
    private static final MoveBuilder BLACK_BISHOP_MOVE_BUILDER = new SliderMoveBuilder(new Direction[] { Direction.NORTH_EAST, Direction.NORTH_WEST, Direction.SOUTH_EAST, Direction.SOUTH_WEST }, false);
    private static final MoveBuilder BLACK_KNIGHT_MOVE_BUILDER = new KnightMoveBuilder(false);
    private static final MoveBuilder BLACK_PAWN_MOVE_BUILDER = new PawnMoveBuilder(false);

    private static final MoveBuilder[] MOVE_BUILDERS;

    static {
        MOVE_BUILDERS = new MoveBuilder[Piece.values().length];
        MOVE_BUILDERS[Piece.WHITE_KING.ordinal()] = WHITE_KING_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.WHITE_QUEEN.ordinal()] = WHITE_QUEEN_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.WHITE_ROOK.ordinal()] = WHITE_ROOK_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.WHITE_BISHOP.ordinal()] = WHITE_BISHOP_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.WHITE_KNIGHT.ordinal()] = WHITE_KNIGHT_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.WHITE_PAWN.ordinal()] = WHITE_PAWN_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.BLACK_KING.ordinal()] = BLACK_KING_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.BLACK_QUEEN.ordinal()] = BLACK_QUEEN_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.BLACK_ROOK.ordinal()] = BLACK_ROOK_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.BLACK_BISHOP.ordinal()] = BLACK_BISHOP_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.BLACK_KNIGHT.ordinal()] = BLACK_KNIGHT_MOVE_BUILDER;
        MOVE_BUILDERS[Piece.BLACK_PAWN.ordinal()] = BLACK_PAWN_MOVE_BUILDER;
    }

    private MoveGenerators() {
    }

    public static MoveBuilder get(Piece piece) {
        return MOVE_BUILDERS[piece.ordinal()];
    }
    
    private enum Direction {
        NORTH(ROW_WIDTH),
        SOUTH(-ROW_WIDTH),
        EAST(+1),
        WEST(-1),
        NORTH_EAST(ROW_WIDTH+1),
        NORTH_WEST(ROW_WIDTH-1),
        SOUTH_EAST(-ROW_WIDTH+1),
        SOUTH_WEST(-ROW_WIDTH-1);

        private final int delta;

        private Direction(int delta) {
            this.delta = delta;
        }
    
        public int getDelta() {
            return delta;
        }
    }
    
    private abstract static class BasicMoveBuilder implements MoveBuilder {
        private final int[] deltas;
        private final boolean isWhite;

        protected BasicMoveBuilder(int[] deltas, boolean isWhite) {
            this.deltas = deltas;
            this.isWhite = isWhite;
        }

        @Override
        public <T> void build(List<T> moves, Explorable board, int from, MoveConstructor<T> moveBuilder) {
            for (int delta : deltas) {
                int to = from + delta;
                Capturable piece = board.getCapturable(to);
                if (piece == null || piece.canBeCapturedBy(isWhite)) {
                    moves.add(moveBuilder.create(from, to));
                }
            }
        }
    }

    private static class KingMoveBuilder extends BasicMoveBuilder {
        public KingMoveBuilder(boolean isWhite) {
            super(new int[] { Direction.NORTH.getDelta(), Direction.SOUTH.getDelta(), Direction.EAST.getDelta(), Direction.WEST.getDelta(), Direction.NORTH_EAST.getDelta(), Direction.NORTH_WEST.getDelta(), Direction.SOUTH_EAST.getDelta(), Direction.SOUTH_WEST.getDelta() }, isWhite);
        }
    }

    private static class KnightMoveBuilder extends BasicMoveBuilder {
        public KnightMoveBuilder(boolean isWhite) {
            super(getDeltas(), isWhite);
        }
    
        private static int[] getDeltas() {
            final int[] deltas = new int[8];
            deltas[0] = 2*Direction.NORTH.getDelta()+Direction.EAST.getDelta();
            deltas[1] = 2*Direction.NORTH.getDelta()+Direction.WEST.getDelta();
            deltas[2] = 2*Direction.SOUTH.getDelta()+Direction.EAST.getDelta();
            deltas[3] = 2*Direction.SOUTH.getDelta()+Direction.WEST.getDelta();
            deltas[4] = Direction.NORTH.getDelta()+2*Direction.EAST.getDelta();
            deltas[5] = Direction.NORTH.getDelta()+2*Direction.WEST.getDelta();
            deltas[6] = Direction.SOUTH.getDelta()+2*Direction.EAST.getDelta();
            deltas[7] = Direction.SOUTH.getDelta()+2*Direction.WEST.getDelta();
            return deltas;
        }
    }

    /**
     * Generates moves for pawns.
     * <br>WARNING: This is a very basic implementation. It does not manage en passant, promotion.
     */
    private static class PawnMoveBuilder implements MoveBuilder {
        private final boolean isWhite;
        private final int advanceDelta;
        private final int captureDeltaWest;
        private final int captureDeltaEast;
        private final int twoAdvanceRank;
        
        public PawnMoveBuilder(boolean isWhite) {
            this.isWhite = isWhite;
            this.advanceDelta = isWhite ? Direction.NORTH.getDelta() : Direction.SOUTH.getDelta();
            this.captureDeltaWest = isWhite ? Direction.NORTH_WEST.getDelta() : Direction.SOUTH_WEST.getDelta();
            this.captureDeltaEast = isWhite ? Direction.NORTH_EAST.getDelta() : Direction.SOUTH_EAST.getDelta();
            this.twoAdvanceRank = isWhite ? 1 : 6;
        }
        
        @Override
        public <T> void build(List<T> moves, Explorable board, int from, MoveConstructor<T> moveBuilder) {
            int to = from + advanceDelta;
            if (board.getCapturable(to) == null) {
                moves.add(moveBuilder.create(from, to));
                to += advanceDelta;
                if (twoAdvanceRank == board.getRank(from) && board.getCapturable(to)==null) {
                    moves.add(moveBuilder.create(from, to));
                }
            }
            to = from + captureDeltaWest;
            Capturable captured = board.getCapturable(to);
            if (captured != null && captured.canBeCapturedBy(isWhite)) {
                moves.add(moveBuilder.create(from, to));
            }
            to = from + captureDeltaEast;
            captured = board.getCapturable(to);
            if (captured != null && captured.canBeCapturedBy(isWhite)) {
                moves.add(moveBuilder.create(from, to));
            }
        }
    }

    private static class SliderMoveBuilder implements MoveBuilder {
        private final int[] deltas;
        private final boolean isWhite;

        public SliderMoveBuilder(Direction[] deltas, boolean isWhite) {
            this.deltas = Arrays.stream(deltas).mapToInt(Direction::getDelta).toArray();
            this.isWhite = isWhite;
        }

        public <T> void scanDirection(List<T> moves, Explorable board, int from, int delta, MoveConstructor<T> moveBuilder) {
            int to = from + delta;
            while (true) {
                Capturable piece = board.getCapturable(to);
                if (piece == null) {
                    moves.add(moveBuilder.create(from, to));
                } else {
                    if (piece.canBeCapturedBy(isWhite)) {
                        moves.add(moveBuilder.create(from, to));
                    }
                    break;
                }
                to += delta;
            }
        }

        @Override
        public <T> void build(List<T> moves, Explorable board, int from, MoveConstructor<T> moveBuilder) {
            for (int delta : deltas) {
                scanDirection(moves, board, from, delta, moveBuilder);
            }
        }
    }
}

File: hellogo/pkg/chess/MoveGenerators.go
package chess

import (
	"fmt"
)

// ROW_WIDTH is the width of a row on the board.
const ROW_WIDTH = 10

// Direction represents a movement direction with its delta.
type Direction struct {
	delta int
}

func (d Direction) Delta() int {
	return d.delta
}

var (
	NORTH      = Direction{ROW_WIDTH}
	SOUTH      = Direction{-ROW_WIDTH}
	EAST       = Direction{+1}
	WEST       = Direction{-1}
	NORTH_EAST = Direction{ROW_WIDTH + 1}
	NORTH_WEST = Direction{ROW_WIDTH - 1}
	SOUTH_EAST = Direction{-ROW_WIDTH + 1}
	SOUTH_WEST = Direction{-ROW_WIDTH - 1}
)

// MoveGenerators holds move builder singletons.
// TODO This should probably be in a separate file with explicit reference to Move as ChatGPT has a lot of problem with it
type MoveGenerators struct{}

var (
	whiteKingMoveBuilder   = NewKingMoveBuilder[interface{}](true)
	whiteQueenMoveBuilder  = NewSliderMoveBuilder[interface{}]([]Direction{NORTH, SOUTH, EAST, WEST, NORTH_EAST, NORTH_WEST, SOUTH_EAST, SOUTH_WEST}, true)
	whiteRookMoveBuilder   = NewSliderMoveBuilder[interface{}]([]Direction{NORTH, SOUTH, EAST, WEST}, true)
	whiteBishopMoveBuilder = NewSliderMoveBuilder[interface{}]([]Direction{NORTH_EAST, NORTH_WEST, SOUTH_EAST, SOUTH_WEST}, true)
	whiteKnightMoveBuilder = NewKnightMoveBuilder[interface{}](true)
	whitePawnMoveBuilder   = NewPawnMoveBuilder[interface{}](true)

	blackKingMoveBuilder   = NewKingMoveBuilder[interface{}](false)
	blackQueenMoveBuilder  = NewSliderMoveBuilder[interface{}]([]Direction{NORTH, SOUTH, EAST, WEST, NORTH_EAST, NORTH_WEST, SOUTH_EAST, SOUTH_WEST}, false)
	blackRookMoveBuilder   = NewSliderMoveBuilder[interface{}]([]Direction{NORTH, SOUTH, EAST, WEST}, false)
	blackBishopMoveBuilder = NewSliderMoveBuilder[interface{}]([]Direction{NORTH_EAST, NORTH_WEST, SOUTH_EAST, SOUTH_WEST}, false)
	blackKnightMoveBuilder = NewKnightMoveBuilder[interface{}](false)
	blackPawnMoveBuilder   = NewPawnMoveBuilder[interface{}](false)

	moveBuilders []MoveBuilder[interface{}]
)

func init() {
	moveBuilders = make([]MoveBuilder[interface{}], 13) // 13 = number of pieces (including BLOCKER)
	moveBuilders[6] = whiteKingMoveBuilder              // WHITE_KING
	moveBuilders[5] = whiteQueenMoveBuilder             // WHITE_QUEEN
	moveBuilders[4] = whiteRookMoveBuilder              // WHITE_ROOK
	moveBuilders[3] = whiteBishopMoveBuilder            // WHITE_BISHOP
	moveBuilders[2] = whiteKnightMoveBuilder            // WHITE_KNIGHT
	moveBuilders[1] = whitePawnMoveBuilder              // WHITE_PAWN
	moveBuilders[12] = blackKingMoveBuilder             // BLACK_KING
	moveBuilders[11] = blackQueenMoveBuilder            // BLACK_QUEEN
	moveBuilders[10] = blackRookMoveBuilder             // BLACK_ROOK
	moveBuilders[9] = blackBishopMoveBuilder            // BLACK_BISHOP
	moveBuilders[8] = blackKnightMoveBuilder            // BLACK_KNIGHT
	moveBuilders[7] = blackPawnMoveBuilder              // BLACK_PAWN
	// BLOCKER (0) left as nil
}

// GetMoveBuilder returns the generic MoveBuilder[interface{}] for a given Piece.
func GetMoveBuilder(piece *Piece) MoveBuilder[interface{}] {
	// The ordinal in Java is the index in the enum, which matches the order in Piece.go
	// We use the same order for moveBuilders.
	switch piece {
	case nil:
		return nil
	}
	// Find index by matching code and color
	switch piece.Code {
	case 'K':
		return moveBuilders[6]
	case 'Q':
		return moveBuilders[5]
	case 'R':
		return moveBuilders[4]
	case 'B':
		return moveBuilders[3]
	case 'N':
		return moveBuilders[2]
	case 'P':
		return moveBuilders[1]
	case 'k':
		return moveBuilders[12]
	case 'q':
		return moveBuilders[11]
	case 'r':
		return moveBuilders[10]
	case 'b':
		return moveBuilders[9]
	case 'n':
		return moveBuilders[8]
	case 'p':
		return moveBuilders[7]
	default:
		return nil
	}
}

// --- MoveBuilder implementations using generics ---

type BasicMoveBuilder[T any] struct {
	deltas  []int
	isWhite bool
}

func NewBasicMoveBuilder[T any](deltas []int, isWhite bool) BasicMoveBuilder[T] {
	return BasicMoveBuilder[T]{deltas: deltas, isWhite: isWhite}
}

func (b BasicMoveBuilder[T]) Build(moves *[]T, board Explorable, from int, moveBuilder MoveConstructor[T]) {
	for _, delta := range b.deltas {
		to := from + delta
		piece := board.GetCapturable(to)
		if piece == nil || piece.CanBeCapturedBy(b.isWhite) {
			*moves = append(*moves, moveBuilder.Create(from, to))
		}
	}
}

type KingMoveBuilder[T any] struct {
	BasicMoveBuilder[T]
}

func NewKingMoveBuilder[T any](isWhite bool) KingMoveBuilder[T] {
	deltas := []int{
		NORTH.Delta(), SOUTH.Delta(), EAST.Delta(), WEST.Delta(),
		NORTH_EAST.Delta(), NORTH_WEST.Delta(), SOUTH_EAST.Delta(), SOUTH_WEST.Delta(),
	}
	return KingMoveBuilder[T]{NewBasicMoveBuilder[T](deltas, isWhite)}
}

type KnightMoveBuilder[T any] struct {
	BasicMoveBuilder[T]
}

func NewKnightMoveBuilder[T any](isWhite bool) KnightMoveBuilder[T] {
	deltas := []int{
		2*NORTH.Delta() + EAST.Delta(),
		2*NORTH.Delta() + WEST.Delta(),
		2*SOUTH.Delta() + EAST.Delta(),
		2*SOUTH.Delta() + WEST.Delta(),
		NORTH.Delta() + 2*EAST.Delta(),
		NORTH.Delta() + 2*WEST.Delta(),
		SOUTH.Delta() + 2*EAST.Delta(),
		SOUTH.Delta() + 2*WEST.Delta(),
	}
	return KnightMoveBuilder[T]{NewBasicMoveBuilder[T](deltas, isWhite)}
}

type PawnMoveBuilder[T any] struct {
	isWhite        bool
	advanceDelta   int
	captureDeltaW  int
	captureDeltaE  int
	twoAdvanceRank int
}

func NewPawnMoveBuilder[T any](isWhite bool) PawnMoveBuilder[T] {
	advanceDelta := NORTH.Delta()
	captureDeltaW := NORTH_WEST.Delta()
	captureDeltaE := NORTH_EAST.Delta()
	twoAdvanceRank := 1
	if !isWhite {
		advanceDelta = SOUTH.Delta()
		captureDeltaW = SOUTH_WEST.Delta()
		captureDeltaE = SOUTH_EAST.Delta()
		twoAdvanceRank = 6
	}
	return PawnMoveBuilder[T]{
		isWhite:        isWhite,
		advanceDelta:   advanceDelta,
		captureDeltaW:  captureDeltaW,
		captureDeltaE:  captureDeltaE,
		twoAdvanceRank: twoAdvanceRank,
	}
}

func (p PawnMoveBuilder[T]) Build(moves *[]T, board Explorable, from int, moveBuilder MoveConstructor[T]) {
	fmt.Printf("TRACE: PawnMoveBuilder.Build called (isWhite=%v, from=%d)\n", p.isWhite, from)
	to := from + p.advanceDelta
	if board.GetCapturable(to) == nil {
		*moves = append(*moves, moveBuilder.Create(from, to))
		to2 := to + p.advanceDelta
		if p.twoAdvanceRank == board.GetRank(from) && board.GetCapturable(to2) == nil {
			*moves = append(*moves, moveBuilder.Create(from, to2))
		}
	}
	toW := from + p.captureDeltaW
	captured := board.GetCapturable(toW)
	if captured != nil && captured.CanBeCapturedBy(p.isWhite) {
		*moves = append(*moves, moveBuilder.Create(from, toW))
	}
	toE := from + p.captureDeltaE
	captured = board.GetCapturable(toE)
	if captured != nil && captured.CanBeCapturedBy(p.isWhite) {
		*moves = append(*moves, moveBuilder.Create(from, toE))
	}
}

type SliderMoveBuilder[T any] struct {
	deltas  []int
	isWhite bool
}

func NewSliderMoveBuilder[T any](directions []Direction, isWhite bool) SliderMoveBuilder[T] {
	deltas := make([]int, len(directions))
	for i, d := range directions {
		deltas[i] = d.Delta()
	}
	return SliderMoveBuilder[T]{deltas: deltas, isWhite: isWhite}
}

func (s SliderMoveBuilder[T]) scanDirection(moves *[]T, board Explorable, from int, delta int, moveBuilder MoveConstructor[T]) {
	to := from + delta
	for {
		piece := board.GetCapturable(to)
		if piece == nil {
			*moves = append(*moves, moveBuilder.Create(from, to))
		} else {
			if piece.CanBeCapturedBy(s.isWhite) {
				*moves = append(*moves, moveBuilder.Create(from, to))
			}
			break
		}
		to += delta
	}
}

func (s SliderMoveBuilder[T]) Build(moves *[]T, board Explorable, from int, moveBuilder MoveConstructor[T]) {
	for _, delta := range s.deltas {
		s.scanDirection(moves, board, from, delta, moveBuilder)
	}
}
